app/src/main/java/fi/thl/koronahaavi/data/DefaultExposureRepository.kt

```diff
@@ -1,43 +1,89 @@
 
-    override suspend fun saveExposure(exposure: Exposure)
-            = exposureDao.insert(exposure)
+    override suspend fun saveExposure(exposure: Exposure) = exposureDao.insert(exposure)
 
-    override suspend fun getExposure(id: Long)
-            = exposureDao.get(id)
+    override suspend fun getExposure(id: Long) = exposureDao.get(id)
 
-    override suspend fun getAllExposures(): List<Exposure>
-            = exposureDao.getAll()
+    override suspend fun getAllExposures(): List<Exposure> = exposureDao.getAll()
 
-    override suspend fun deleteAllExposures()
-            = exposureDao.deleteAll()
+    override suspend fun deleteAllExposures() = exposureDao.deleteAll()
 
-    override suspend fun deleteExposure(id: Long)
-            = exposureDao.delete(id)
+    override suspend fun deleteExposure(id: Long) = exposureDao.delete(id)
 
-    override fun flowAllExposures(): Flow<List<Exposure>>
-            = exposureDao.flowAll()
+    override suspend fun deleteExpiredExposuresAndTokens() {
+        val expireTime = getExpireTime()
 
-    override fun flowHasExposures(): Flow<Boolean>
-            = exposureDao.flowCount().map { it > 0 }
+        getAllExposures().forEach {
+            if (it.detectedDate.isBefore(expireTime)) {
+                Timber.d("Deleting exposure detected ${it.detectedDate}")
+                deleteExposure(it.id)
+            }
+        }
+
+        getAllKeyGroupTokens().forEach {
+            if (it.updatedDate.isBefore(expireTime)) {
+                Timber.d("Deleting key group updated ${it.updatedDate}")
+                deleteKeyGroupToken(it)
+            }
+        }
+    }
+
+    override fun flowAllExposures(): Flow<List<Exposure>> = exposureDao.flowAll()
+
+    override fun flowExposureNotifications(): Flow<List<ExposureNotification>> =
+        exposureDao.flowAll().map { all ->
+            // First filter out expired exposures, then group by created date, but because
+            // created date could theoretically be on a different second, group manually
+            // instead of using created date directly as a map key.
+
+            val result = mutableMapOf<ZonedDateTime, List<Exposure>>()
+            val expireTime = getExpireTime()
+
+            all.filter { e ->
+                e.detectedDate.isAfter(expireTime)
+            }.forEach { e ->
+                // find existing key that is close enough, or create new key
+                val key = result.keys.find {
+                    Duration.between(it, e.createdDate).abs().seconds < EXPOSURE_GROUPING_THRESHOLD_SECONDS
+                } ?: e.createdDate
+
+                // append to existing, or create a new entry
+                result[key] = result.getOrDefault(key, listOf()).plus(e)
+            }
+
+            result.map { ExposureNotification(it.key, it.value) }
+        }
+
+    private fun getExpireTime() = ZonedDateTime.now().minusDays(EXPOSURE_VALID_SINCE_DETECTION_DAYS.toLong())
```

Comment:
SettingsRepository has exposureValidDays, but it isn't used here - should it or does that and EXPOSURE_VALID_SINCE_DETECTION_DAYS have different meanings?

Reply:
Yes should be used here, added it